File: deck/base/code/native/rust/runtime/calendar.rs
mod calendar {
    use chrono::{DateTime, Utc, TimeZone, Datelike, Timelike, Months};
    fn at(millis: i64) -> DateTime<Utc> {
        Utc.timestamp_millis_opt(millis).single().unwrap_or_else(|| Utc.timestamp_opt(0, 0).single().unwrap())
    }
    pub fn format(millis: i64) -> String { at(millis).format("%Y-%m-%dT%H:%M:%S%.3fZ").to_string() }
    pub fn parse(text: String) -> i64 {
        DateTime::parse_from_rfc3339(&text).map(|date| date.timestamp_millis()).unwrap_or(0)
    }
    pub fn year(millis: i64) -> i64 { at(millis).year() as i64 }
    pub fn month(millis: i64) -> i64 { at(millis).month() as i64 }
    pub fn day(millis: i64) -> i64 { at(millis).day() as i64 }
    pub fn hour(millis: i64) -> i64 { at(millis).hour() as i64 }
    pub fn minute(millis: i64) -> i64 { at(millis).minute() as i64 }
    pub fn second(millis: i64) -> i64 { at(millis).second() as i64 }
    pub fn weekday(millis: i64) -> i64 { at(millis).weekday().num_days_from_sunday() as i64 }
    pub fn build(year: i64, month: i64, day: i64, hour: i64, minute: i64, second: i64) -> i64 {
        Utc.with_ymd_and_hms(year as i32, month as u32, day as u32, hour as u32, minute as u32, second as u32)
            .single()
            .map(|date| date.timestamp_millis())
            .unwrap_or(0)
    }
    pub fn add_months(millis: i64, count: i64) -> i64 {
        let date = at(millis);
        let shifted = if count >= 0 {
            date.checked_add_months(Months::new(count as u32))
        } else {
            date.checked_sub_months(Months::new((-count) as u32))
        };
        shifted.map(|d| d.timestamp_millis()).unwrap_or(millis)
    }
    pub fn add_years(millis: i64, count: i64) -> i64 { add_months(millis, count * 12) }
}

File: deck/seed/code/native/browser/runtime/calendar.ts
// Calendar over the host Date, all in UTC. Timestamps are epoch milliseconds. Formatting uses Date.toISOString
// (always millisecond precision with a Z suffix); parsing uses Date.parse over the same ISO 8601 shape.
const calendar = {
  format: (millis: number): string => new Date(millis).toISOString(),
  parse: (text: string): number => Date.parse(text),
  year: (millis: number): number => new Date(millis).getUTCFullYear(),
  month: (millis: number): number => new Date(millis).getUTCMonth() + 1,
  day: (millis: number): number => new Date(millis).getUTCDate(),
  hour: (millis: number): number => new Date(millis).getUTCHours(),
  minute: (millis: number): number => new Date(millis).getUTCMinutes(),
  second: (millis: number): number => new Date(millis).getUTCSeconds(),
  weekday: (millis: number): number => new Date(millis).getUTCDay(),
  build: (
    year: number,
    month: number,
    day: number,
    hour: number,
    minute: number,
    second: number,
  ): number => Date.UTC(year, month - 1, day, hour, minute, second),
  addMonths: (millis: number, count: number): number => {
    const date = new Date(millis)
    date.setUTCMonth(date.getUTCMonth() + count)
    return date.getTime()
  },
  addYears: (millis: number, count: number): number => {
    const date = new Date(millis)
    date.setUTCFullYear(date.getUTCFullYear() + count)
    return date.getTime()
  },
}
